### ISSUE TITLE
Support dynamic partition in loss function

### ISSUE THREAD

**Issue by jfpuget** [tensorflow/tensorflow#14655]
### System information

Python 3.6.1, Ubuntu 16.04, TF 1.3

### Describe the problem

I am trying to implement a loss function in Tensorflow similar to the Theano loss function described here: https://github.com/Lasagne/Lasagne/issues/767

I have tried several code but they all lead to runtime error.  Seems that dynamically sized tensors aren't supported in loss function.  I have tried various optimizers, including adam, without success.

When tested standalone, the function works fine and produces results similar to a numpy based implementation.

### Source code / logs
  Here is one code I tried:

    import tensorflow as tf
    
    def pair_loss(y_true, y_pred):
        y_true = tf.cast(y_true, tf.int32)
        parts = tf.dynamic_partition(y_pred, y_true, 2)
        y_pos = parts[1]
        y_neg = parts[0]
        y_pos = tf.expand_dims(y_pos, 0)
        y_neg = tf.expand_dims(y_neg, -1)
        out = tf.sigmoid(y_neg - y_pos)
        return tf.reduce_mean(out, axis=-1)

Here is the theano code for reference:

    import theano
    
    def calc_auroc_loss(pred_vr, y_vr):
        pos_pred_vr = pred_vr[y_vr.nonzero()]
        neg_pred_vr = pred_vr[theano.tensor.eq(y_vr, 0).nonzero()]
        pred_diffs_vr = pos_pred_vr.dimshuffle(0, 'x') - neg_pred_vr.dimshuffle('x', 0)
        num_pairs_vr = theano.tensor.sum(theano.tensor.eq(y_vr, 1)) * theano.tensor.sum(theano.tensor.eq(y_vr, 0))
        auroc_vr = theano.tensor.sum(theano.tensor.nnet.sigmoid(pred_diffs_vr)) / num_pairs_vr
        return -auroc_vr

**Comment by ppwwyyxx** [comment]
You can literally line-by-line translate your theano code. Maybe you missed `tf.boolean_mask`.

**Comment by jfpuget** [comment]
boolean_mask does not work either.

**Comment by ppwwyyxx** [comment]
I guess you need to tell everyone what is "does not work". Post errors if you saw any, and better with some code.

**Comment by jfpuget** [comment]
OK, I will extract code from my full code and post it here.

**Comment by jfpuget** [comment]
I attach a code and its output that shows the issue.  I think the issue comes from the broadcast in line 16 of the attached code.

[tf_bug.py.txt](https://github.com/tensorflow/tensorflow/files/1487749/tf_bug.py.txt)
[tf_bug_output.txt](https://github.com/tensorflow/tensorflow/files/1487751/tf_bug_output.txt)

**Comment by ppwwyyxx** [comment]
It's just another constraint added by keras and not a tensorflow issue. It's complaining that your custom loss function return something of a different shape. Using normal tensorflow optimizer with this custom loss will be fine.

**Comment by jfpuget** [comment]
Not really, because the return shape is the same whether I have this broadcast in line 16 or something without broadcast. Yet the latter does not trigger an issue.  

If I use tf.keras instead of keras, does this qualify better as a tf issue?

**Comment by ppwwyyxx** [comment]
Line 16 in your code is `out = tf.sigmoid(y_neg - y_pos)`. I don't know what you mean by not doing broadcast, since y_neg and y_pos may have different shape. Maybe you can clarify that by providing a modification that makes it work.
I'm sure gradient can pass your loss function in raw tensorflow, because if you use `tf.reduce_mean(out)` it can run. Why it does not work in keras is not something I'm able to help. But you can certainly leave it here for others to answer. At least now the problem is more clear.

**Comment by jfpuget** [comment]
Your previous comment made me find the issue, thanks a lot.  The issue was in my code indeed.  I should have used tf.reduce_mean(out) instead of tf.reduce_mean(out, axis=-1).

**Comment by ppwwyyxx** [comment]
Oh great. I thought you want `axis=-1` on purpose, because Keras does allow you to return a vector in the loss function.

**Comment by jfpuget** [comment]
Yes, Keras allows this, which is why the error was a bit obfuscated for me.  Thanks again for your help, and sorry for the distraction.

**Comment by jfpuget** [comment]
This issue still appears in the open issues because of the tag awaiting response, but there is no more response to provide.  It can be closed.

**Comment by ashimb9** [comment]
@jfpuget I tried using the corrected version of your pair_loss function above on a GRU model. It works fine for about 2 seconds after which I keep getting `loss: nan` during the remainder of the training. Curiously enough, the accuracy keeps increasing. Would you happen to know why it might be returning `nan`?

**Comment by jfpuget** [comment]
Hi,  some of your gradient is exploding.  Try using a different optimizer, or try clipping gradient.  The net result is that some predictions are leading to overflow in computation, and therefore nan, but it seems that in most cases your predictions are getting better.

**Comment by ashimb9** [comment]
@jfpuget Thank you for responding. Gradient clipping and optimizer selection only delayed the nanification process, but could not prevent it. However, based on the discussion [here](https://github.com/keras-team/keras/issues/2134), simply increasing the batch size got it working again.

**Comment by cweill** [comment]
@ashimb9: You might be getting Nans when your mini-batch contains only a single class. I'm not sure of that function's behavior if either of the partitions is empty, or what a good default behavior would be. I don't believe that behavior is defined for RankNet or LambdaRank, where the assumption is that you are dealing with labeled pairs.
